homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,717 @@
+        """Volume level of the media player (0..1)."""
+        # pylint: disable=arguments-differ
+        value = state_variable.value
+        if value is None:
+            _LOGGER.debug('%s.volume_level(): Got no value', self)
+            return None
+
+        max_value = state_variable.max_value or 100
+        return min(value / max_value, 1.0)
+
+    @requires_action('RC', 'SetVolume')
+    async def async_set_volume_level(self, action, volume):
+        """Set volume level, range 0..1."""
+        # pylint: disable=arguments-differ
+        argument = action.argument('DesiredVolume')
+        state_variable = argument.related_state_variable
+        min_ = state_variable.min_value or 0
+        max_ = state_variable.max_value or 100
+        desired_volume = int(min_ + volume * (max_ - min_))
+
+        await action.async_call(InstanceID=0,
+                                Channel='Master',
+                                DesiredVolume=desired_volume)
+
+    @property
+    @requires_state_variable('RC', 'Mute')
+    def is_volume_muted(self, state_variable):
+        """Boolean if volume is currently muted."""
+        # pylint: disable=arguments-differ
+        value = state_variable.value
+        if value is None:
+            _LOGGER.debug('%s.is_volume_muted(): Got no value', self)
+            return None
+
+        return value
+
+    @requires_action('RC', 'SetMute')
+    async def async_mute_volume(self, action, mute):
+        """Mute the volume."""
+        # pylint: disable=arguments-differ
+        desired_mute = bool(mute)
+        await action.async_call(InstanceID=0,
+                                Channel='Master',
+                                DesiredMute=desired_mute)
+
+    @requires_action('AVT', 'Pause')
+    async def async_media_pause(self, action):
+        """Send pause command."""
+        # pylint: disable=arguments-differ
+        await action.async_call(InstanceID=0)
+
+    @requires_action('AVT', 'Play')
+    async def async_media_play(self, action):
+        """Send play command."""
+        # pylint: disable=arguments-differ
+        await action.async_call(InstanceID=0, Speed='1')
+
+    @requires_action('AVT', 'Stop')
+    async def async_media_stop(self, action):
+        """Send stop command."""
+        # pylint: disable=arguments-differ
+        await action.async_call(InstanceID=0)
+
+    @requires_action('AVT', 'SetAVTransportURI')
+    async def async_play_media(self, action, media_type, media_id, **kwargs):
+        """Play a piece of media."""
+        # pylint: disable=arguments-differ
+
+        # queue media
+        meta_data = await self._construct_play_media_metadata(media_type,
+                                                              media_id)
+        await action.async_call(InstanceID=0,
+                                CurrentURI=media_id,
+                                CurrentURIMetaData=meta_data)
+        await asyncio.sleep(0.25)
```

Comment:
Why sleep here? Making sure the file gets queued before calling play? Is 0.25 guaranteed to be enough for that?

Reply:
My DLNA-enabled devices gave problems when the sleep was not added. I cannot guarantee if 0.25 second will be enough.

Another option might be spin-waiting for a state variable change. Would that be ok? Also, how long is the component allowed to wait?

Reply:
With spin-waiting I meant waiting, up to a maximum time, checking if a given state variable has been changed via the notification system.

Reply:
Yeah, that upnp api sounds really ugly on this regard, but I suppose the sleep (or wait for a metadata notification change) is required to set it into the mode where it can play the new file, and thus it really depends on the device if 0.25 is enough or not.

I'm unsure what's the nicest way to go here, tbh, so I hope someone else will chime in here. As it's asyncio.sleep, it won't be blocking anything else so making the sleep time higher would not be a problem (but it hampers the user experience, of course). I would go to the route of "spin-waiting" for a specific time, as for properly working devices that won't be an issue (+ it will wait no longer than required).